pkg/overlay/cache.go

```diff
@@ -107,51 +109,83 @@ func (o *Cache) Bootstrap(ctx context.Context) error {
 			return err
 		}
 	}
-	// called after kademlia is bootstrapped
-	// needs to take RoutingTable and start to persist it into the cache
-	// take bootstrap node
-	// get their route table
-	// loop through nodes in RT and get THEIR route table
-	// keep going forever and ever
-
-	// Other Possibilities: Randomly generate node ID's to ask for?
-
-	_, err = o.DHT.GetRoutingTable(ctx)
 
 	return err
 }
 
-// Refresh walks the network looking for new nodes and pings existing nodes to eliminate stale addresses
+// Refresh updates the cache db with the current DHT.
+// We currently do not penalize nodes that are unresponsive,
+// but should in the future.
 func (o *Cache) Refresh(ctx context.Context) error {
-	// iterate over all nodes
-	// compare responses to find new nodes
-	// listen for responses from existing nodes
-	// if no response from existing, then mark it as offline for time period
-	// if responds, it refreshes in DHT
-	_, rtErr := o.DHT.GetRoutingTable(ctx)
+	log.Print("starting cache refresh")
+	table, err := o.DHT.GetRoutingTable(ctx)
+	if err != nil {
+		zap.Error(OverlayError.New("Error getting routing table", err))
+		return err
+	}
+
+	buckets, _ := table.GetBuckets()
 
-	if rtErr != nil {
-		return rtErr
+	for _, b := range buckets {
+		log.Printf("refreshing bucket %+v\n", b)
+		bucketNodes := b.Nodes()
+		for _, bn := range bucketNodes {
+			log.Printf("bucket node %+v\n", bn)
+		}
 	}
 
-	_, err := o.DHT.GetNodes(ctx, "0", 128)
+	nodes, err := o.DHT.GetNodes(ctx, "", 1024)
+
+	/// kick off concurrent refreshes
+	randomID, err := kademlia.NewID()
 
 	if err != nil {
-		return err
+		zap.Error(OverlayError.New("Error generating random ID", err))
+		log.Printf("error finding random ID %+v\n", err)
 	}
 
-	return nil
+	nn, err := o.DHT.FindNearNodes(ctx, randomID)
+	if err != nil {
+		zap.Error(OverlayError.New("Error finding node", err))
+	}
+
+	fmt.Printf("found near nodes %+v\n", nn)
+
+	for _, n := range nn {
+		log.Printf("getting routing table for node %+v\n", n)
+		t, err := kademlia.GetNodeRoutingTable(ctx, kademlia.NodeID(n.Id))
+		if err != nil {
+			zap.Error(OverlayError.New("Error getting node routing table", err))
+			log.Printf("error getting routing table from foreign node %+v\n", err)
+		}
+		log.Printf("got routing table back %+v\n", t)
+	}
+
+	for _, node := range nodes {
```

Comment:
we're looping through the 1024 nodes we requested. is there a way to do a wider for loop so we continue to request more nodes?